File: 5/solution.js
#!/usr/bin/env node

const fs = require('fs')
const path = require('path')

const debug = process.env.DEBUG == 1 || false

fs.readFile(path.join(__dirname, '/input'), 'utf8', (err, data) => {
    const lines = data.split('\n')
    let max = 0
    let seated = []
    let possibleSeats = []

    for (const line of lines) {
      const binRow = line.slice(0, 7)
      const binCol = line.slice(7)

      const row = parseBin(binRow)
      const col = parseBin(binCol)
      const seatId = row * 8 + col
      debug && console.log(`${seatId}: ${row}, ${col}`)

      seated.push(seatId)

      max = seatId > max ? seatId : max
    }

    console.log(`Highest seat ID: ${max}`)

    let preSeat
    let maybeMySeat = seated.includes(8)
    let postSeat = seated.includes(9)

    const uLimit = 127 * 8
    for (i = 10; i < uLimit; i++) {
      preSeat = maybeMySeat
      maybeMySeat = postSeat
      postSeat = seated.includes(i)

      if (preSeat && postSeat && !maybeMySeat) {
        possibleSeats.push(i - 1)
      }
    }

    console.log(`Possible seats: ${possibleSeats}`)
})

function parseBin(str) {
  let res = 0
  const strLen = str.length
  for (i = 0; i < strLen; i++) {
    const char = str.charAt(strLen - i - 1)
    if (char === 'R' || char === 'B') {
      res += 1 << i
    }
  }

  return res
}

File: 5/solution.rs
use std::env;
use std::fs;

fn main() {
    let args: Vec<String> = env::args().collect();
    let filename = &args[1];

    let seats: Vec<i32> = fs::read_to_string(filename)
        .expect("Unable to read input file")
        .lines()
        .map(|line| seat_id(&line))
        .collect();

    println!("Max ID: {}", seats.iter().max().unwrap());

    let l_limit: i32 = 10;
    let u_limit: i32 = 127 * 8;
    let mut prev: bool;
    let mut maybe: bool = seats.contains(&(l_limit - 2));
    let mut post: bool = seats.contains(&(l_limit - 1));
    let mut possible_seats = vec![];

    for i in l_limit..u_limit {
        prev = maybe;
        maybe = post;
        post = seats.contains(&i);
        if prev && post & !maybe {
            possible_seats.push(i - 1);
        }
    }

    println!("Possible seats: {:?}", possible_seats);
}

fn seat_id(s: &str) -> i32 {
    (parse_bin(&s[..7]) * 8) + parse_bin(&s[7..])
}

fn parse_bin(s: &str) -> i32 {
    let len = s.len();
    let mut val = 0;

    for (i, c) in s.chars().enumerate() {
        if c == 'R' || c == 'B' {
            val += 1 << (len - i - 1);
        }
    }

    val
}